src/TransponderCode.hpp

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The XCSoar Project
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+#include <cstdlib>
+#include <string>
+
+/**
+ * This class stores a transponder code.
+ * It accepts the code "as is" and converts it to an octal (base 8) 
+ * representation of the code. The output needs to be formatted with %04o to
+ * get a usable transponder code.
+ */
+class TransponderCode {
+  uint_least16_t value;
+
+public:
+  /**
+   * Uninitialized.
+   */
+  TransponderCode() = default;
+
+  /**
+   * Construct an empty instance.  Its IsDefined() method will return
+   * false.
+   */
+  static TransponderCode Null() noexcept {
+    return TransponderCode(9);
+  }
+
+  explicit TransponderCode(uint_least16_t code) noexcept {
+    SetCode(code);
+  }
+
+  explicit TransponderCode(const char* code) noexcept : value(ParseCode(code)) {}
+
+  void SetCode(uint_least16_t code) noexcept {
+    value = ParseCode(std::to_string(code).c_str());
```

Comment:
What the f... is this? This allocates heap memory to hold the decimal representation of the code as a `std::string`, then takes the C string of it and passes it to `ParseCode()` .... this looks as confusing and wrong and error prone as it could be.